### DISCUSSION TITLE
Calculation of stoichiometric / theoretical combustion air

### DISCUSSION THREAD

**Discussion by ngom52** [equinor/neqsim#806] [use of NeqSim] [answered]
Hi everyone,

I'm wondering if there is a Python function/method in NeqSim for calculation of stoichiometric / theoretical combustion air; as NeqSim is able to calculate the calorific value of natural gas mixtures, I'm assuming that there is a calculation for stoichiometric air somewhere in the background (although I have not yet found any from the examples in the wiki)

Just a short background about me: I am an engineer in a power utility company in Malaysia with a background in Mechanical Engineering, therefore my knowledge of chemistry is quite rudimentary.
I am currently trying to create a Python based thermodynamic model for one of my company's single shaft combined cycle plant. I was able to use NeqSim to calculate the fuel gas CV based on the example in the wiki, with pretty good accuracy (about 0.37% difference compared with the data from our gas supplier)

In order to complete the model, I need to be able to calculate the stoichiometric air; would really appreciate if anyone can share an example of how to do this in Python. Thanks in advance for your help!

**Comment by EvenSol** [comment, marked as answer]
Hi.

In neqsim (2.5.5+) you can calculate afr using a script like:

 ```
elements_h = 0.0
elements_c = 0.0
sum_hc = 0.0
molmass_hc = 0.0
wtfrac_hc = 0.0

for i in range(fluid.getNumberOfComponents()):
    if fluid.getComponent(i).isHydrocarbon():
        sum_hc = sum_hc + fluid.getComponent(i).getz()
        molmass_hc = molmass_hc + fluid.getComponent(i).getz() * fluid.getComponent(i).getMolarMass()
        elements_c = elements_c + fluid.getComponent(i).getz() * fluid.getComponent(i).getElements().getNumberOfElements("C")
        elements_h = elements_h + fluid.getComponent(i).getz() * fluid.getComponent(i).getElements().getNumberOfElements("H")

if sum_hc == 0:
    print('no hydrocarbons in fluid ')
else:
    wtfrac_hc = molmass_hc / fluid.getMolarMass()
    molmass_hc /= sum_hc
    elements_c /= sum_hc
    elements_h /= sum_hc
    aconst = elements_c + elements_h / 4
    afr = aconst * (32.0 + 3.76 * 28.0) / 1000.0 / molmass_hc * wtfrac_hc
    print('ideal air fluid ratio ', afr)
```

**Reply by ngom52** [reply]
Hi, many thanks for your solution. I have adapted it to my program and it works wonderfully 👍

**Reply by ngom52** [reply]
Hi just a follow up (probably dumb) question, I wish to clarify about the .getz() method; does the Z here represent the compressibility factor, or stoichiometric coefficient for the oxidizer?

**Reply by EvenSol** [reply]
Hi. 

The method is a version of what you find described [here](https://www.omnicalculator.com/chemistry/air-fuel-ratio-afr).

The fluid.getComponent(i).getz() is the overall mole fraction of component i in the fluid. 

The method fluid.getPhase('gas').getComponent(i).getx() will return the mole fraction of a component i in the gas phase.

The method fluid.getPhase('gas').getZ()  will return the compressibility factor for the gas phase.

**Reply by ngom52** [reply]
Hi, thanks for the clarification!
